### PR TITLE
populated data from ordapi

### PR DESCRIPTION
**Pull request by samMcCarthyBJSS**
Story [AB#8119](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/8119)
Task [AB#7856](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/7856)

### CODE REVIEW

File: app/pages/sections/catalogue-solutions/routes.js [hunk 1]
```diff
@@ -49,19 +49,30 @@ export const catalogueSolutionsRoutes = (authProvider, addContext, sessionManage
...
+        'deliveryDate-year': date[0],
+        'deliveryDate-month': date[1],
+        'deliveryDate-day': date[2],
+        quantity: selectedPrice.quantity,
+        selectEstimationPeriod: selectedPrice.estimationPeriod,
+        price: selectedPrice.price,
+      };
+    } else {
+      selectedPrice = await getSelectedPrice({ selectedPriceId, accessToken });
+    }
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/routes.js:73]: I think we can tidy this up a little. Let's add a function in the controller called something like `getFormData({ orderId, orderItemId, selectedPriceId, accessToken })` and we can move this logic there. This way this function will work out whether to get this data from ordapi or bapi

File: app/pages/sections/catalogue-solutions/order-item/controller.js [hunk 2]
```diff
@@ -77,15 +77,14 @@ export const getOrderItemContext = async ({
...
 }) => {
   const selectedPriceManifest = getSelectedPriceManifest({
     provisioningType: selectedPrice.provisioningType,
     type: selectedPrice.type,
   });
 
-  const populatedData = {
-    price: selectedPrice.price,
-  };
+  const populatedData = formData !== undefined ? formData : { price: selectedPrice.price };
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/controller.js:87]: we may not need to do this if we create that new function to getFormData.

**Comment by alibombaye** [comment]
we are going to need some UI tests in flat/ondemand called withSavedData.ui.test.js that nock ordapi and check it populates all the inputs with the data provided

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+  serviceRecipient: {
+    odsCode: 'OX3',
+    name: 'Some service recipient 2',
+  },
+  catalogueItemType: 'Solution',
+  catalogueItemName: 'Some catalogue name',
+  catalogueItemId: '10000-001',
+  deliveryDate: '2020-04-27',
+  quantity: 3,
+  estimationPeriod: 'perMonth',
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:44]: this should be 'month'

Comment (samMcCarthyBJSS) [reply app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:44]: it shoudln't that doesn't match up to what is in the manifest so it doesn't autopopulate

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  .page('http://localhost:1234/order/some-fake-page')
+  .afterEach(async (t) => {
+    const isDone = nock.isDone();
+    if (!isDone) {
+      nock.cleanAll();
+    }
+
+    await t.expect(isDone).ok('Not all nock interceptors were used!');
+  });
+
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:90]: Can we check the title is rendered with the data from ordapi please

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    await selectedRecipientIdState();
+    if (postRoute) {
+      await solutionNameState();
+      await serviceRecipientNameState();
+      await selectedPriceState(orderItem);
+    }
+  }
+};
+
+fixture('Catalogue-solutions - flat ondemand - general')
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:80]: change general to withSavedData

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    description: 'per consultation',
+  },
+  price: 0.1,
+};
+
+const selectedPriceState = ClientFunction((orderItemState) => {
+  const cookieValue = JSON.stringify(orderItemState);
+
+  document.cookie = `selectedPrice=${cookieValue}`;
+});
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:59]: Think we can safely remove some of these state cookies now
solutionNameState
selectedRecipientIdState
serviceRecipientNameState
selectedPriceState

Comment (samMcCarthyBJSS) [reply app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:59]: I removed the ones we don't need but still need some for the post

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+test('should render the price table content', async (t) => {
+  await pageSetup();
+  await t.navigateTo(pageUrl);
+
+  const priceInput = Selector('[data-test-id="question-price"] input');
+
+  await t
+    .expect(priceInput.exists).ok()
+    .expect(priceInput.getAttribute('value')).eql('0.1');
+});
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:140]: can you also add an expectation to this test that is renders `per consultation` as the unit of order

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js (hunk 6 above)
Comment (alibombaye) [reply app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:59]: mmm ok leave them in.. I think we might want to save in session the orderItem we get back from ordapi. we'll do that as part of the next task 👍

Comment (samMcCarthyBJSS) [reply app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:59]: got that covered it saves it as selectedPrice though

Comment (alibombaye) [reply app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:59]: ah nice ok just remove the ones we don't need then. nice work

File: app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+test('should render the title', async (t) => {
+  await pageSetup();
+  await t.navigateTo(pageUrl);
+
+  const title = Selector('h1[data-test-id="order-item-page-title"]');
+
+  await t
+    .expect(title.exists).ok()
+    .expect(await extractInnerText(title)).eql('Some catalogue name information for Some service recipient 2 (OX3)');
```
Comment (alibombaye) [inline app/pages/sections/catalogue-solutions/order-item/flat/ondemand/ui-tests/withSavedData.ui.test.js:99]: 👍

**Pull request merged**